### DISCUSSION TITLE
Is there a way to view population diversity statistics?

### DISCUSSION THREAD

**Discussion by Jan-Loettgen** [PasaOpasen/geneticalgorithm2-6.8.7#22] [Show and tell]
Hi,

I am really loving the package.

I believe I am having issues with premature population convergence.
Is there a way view some population statistics throughout the generations? for example the number of unique pops, or the gene diversity amongst the population.

Thanks

**Comment by PasaOpasen** [comment]
@Jan-Loettgen do u mean to view in real time like progress bar?
if no, u can create some callback to collect these data and plot it each 10-50 generations. Is it what u want?

**Reply by Jan-Loettgen** [reply]
I was looking for something already implemented to look at population diversity after running the genetical algorithm, so not like a progress bar. I will implement something using callbacks, thanks.

**Reply by PasaOpasen** [reply]
@Jan-Loettgen did u? can u show what exactly u wanted to see?

**Reply by Jan-Loettgen** [reply]
Not yet I haven't, but here is a screenshot of what I am looking for. This is from my own implementation of a GA, but yours runs better.
![GA_diversity](https://user-images.githubusercontent.com/70091840/124362133-7c319880-dc33-11eb-9886-d3655fa545d8.PNG)

**Reply by PasaOpasen** [reply]
@Jan-Loettgen what is gene diversity? what does it mean?

**Reply by Jan-Loettgen** [reply]
There are a few different ways to measure gene diversity. Essentially it tries to measure the similarity among the population, i.e. two genomes can be similar without being identical. I will add the code for my implementation, but if you do some research there are a few ways to measure population diversity.
Essentially call this for every generation:
    gene_diversity = 0
    for index in range(population.shape[0]):
      gene_diversity += np.count_nonzero(population[index,:]!=population[index:,:])

**Comment by PasaOpasen** [comment]
@Jan-Loettgen I made new callback

take a look on [file](https://github.com/PasaOpasen/geneticalgorithm2/blob/master/tests/plot_diversities.py)
